### PR TITLE
Event system implementation

### LINKED ISSUE #326: Introduce Event System
**Describe what you want implemented** 
Contracts run by the virtual machine should be able to call a new host function, allowing them to emit a log of a certain event. A list of these logs should then be returned as a result of a transaction execution, in order of their emission.

The event logs should contain at least the following information:

```text
struct EventLog {
    contract_id,
    data,
}
```

where `contract_id` is the ID of the contract that emitted the event, and `data` is the data that the contract has passed in this event.

**Describe "Why" this is needed**
An event system may be used upstream for a myriad of use cases. They can serve as a logging mechanism, as pub-sub topics for taking action on specific data, etc...

**Describe alternatives you've considered**
N/A

**Additional context**
Upstream, event logs are stored as part of the blockchain. Their inclusion can be proven by using a structure similar to [Bloom filter](https://en.wikipedia.org/wiki/Bloom_filter).

### PR DESCRIPTION
**Pull request by ureeves**
This PR implements an event system. It works by introducing a new host import - named `emit` - that when called pushes an event onto a list. The list is subsequently returned - wrapped in a `Receipt` - together with the result of the call.

Each event contains the following data:

```rust
pub struct Event {
    origin: ContractId, // originating smart contract ID
    name: String,       // name of the type of event
    data: Vec<u8>,      // arbitrary data emitted with the event
}
```

The smart contract engineer has the option of passing any structure annotated with the `rusk_uplink::event` attribute, and implementing the `rusk_uplink::Event` trait.

Resolves #326

### CODE REVIEW

File: rusk-uplink/src/definitions.rs [hunk 1]
```diff
@@ -274,6 +278,41 @@ impl<'a> RawTransaction<'a> {
     }
 }
 
+#[derive(Debug, Default)]
+pub struct RawEvent<'a> {
```
Comment (krl) [inline rusk-uplink/src/definitions.rs:282]: Either make data a slice with a lifetime, or make name a string, having both makes little sense

File: rusk-uplink/src/definitions.rs [hunk 2]
```diff
@@ -102,6 +102,10 @@ pub trait Transaction: Archive {
     type Return;
 }
 
+pub trait Event: Archive {
```
Comment (krl) [inline rusk-uplink/src/definitions.rs:105]: Consider removing this trait, as discussed.

File: rusk-uplink/src/definitions.rs (hunk 1 above)
Comment (krl) [reply rusk-uplink/src/definitions.rs:282]: Or maybe even use a 'static str for the name

File: src/state.rs [hunk 3]
```diff
@@ -28,6 +28,19 @@ use crate::modules::ModuleConfig;
 use crate::modules::{compile_module, HostModules};
 use crate::{Schedule, VMError};
 
+#[derive(Debug, Clone)]
```
Comment (krl) [inline src/state.rs:31]: Why are you using pub fields here? It seems like allowing runtime modification of the name or origin should not be allowed.

File: src/state.rs [hunk 4]
```diff
@@ -28,6 +28,19 @@ use crate::modules::ModuleConfig;
...
+pub struct Event {
+    pub origin: ContractId,
+    pub name: String,
+    pub data: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Receipt<R> {
+    pub ret: R,
```
Comment (krl) [inline src/state.rs:40]: Instead of having a public `pub` field, implement `Deref<Target = R>` for Receipt.

**Review by krl — CHANGES_REQUESTED** [review]
Looks good in general, but please avoid public fields in struct unless well motivated.

File: rusk-uplink/src/definitions.rs (hunk 1 above)
Comment (ureeves) [reply rusk-uplink/src/definitions.rs]: Went with taking a `String`.

File: rusk-uplink/src/definitions.rs (hunk 2 above)
Comment (ureeves) [reply rusk-uplink/src/definitions.rs]: Removed

File: src/state.rs (hunk 3 above)
Comment (ureeves) [reply src/state.rs:33]: You're right. Went with "getters" instead.

File: src/state.rs (hunk 4 above)
Comment (ureeves) [reply src/state.rs:63]: Makes a lot of sense. Done

Comment (krl) [reply src/state.rs:63]: Still not done afaics

Comment (krl) [reply src/state.rs:63]: No deref and ret is still public.

**Review by krl — CHANGES_REQUESTED** [review]

**Review by krl — APPROVED** [review]
lgtm!

**Pull request closed without merging**